segment_coverage/segment_coverage/main.py

```diff
@@ -0,0 +1,476 @@
+		# write the pixel array to a temporary file then atomically rename it
+		final_path = os.path.join(self.base_dir, 'coverage-maps',
+				'{}_{}_coverage.png'.format(self.channel, quality))
+		temp_path = final_path.replace('_coverage', '_{}'.format(random.getrandbits(32)))
+		common.ensure_directory(temp_path)
+		matplotlib.image.imsave(temp_path, colours)
+		os.rename(temp_path, final_path)
+		self.logger.info('Coverage map for {} created'.format(quality))
+
+
+	def run(self):
+		"""Loop over available hours for each quality, checking segment coverage."""
+		self.logger.info('Starting')
+
+		while not self.stopping.is_set():
+
+			for quality in self.qualities:
+				if self.stopping.is_set():
+					break
+
+				path = os.path.join(self.base_dir, self.channel, quality)
+				hours = [name for name in os.listdir(path) if not name.startswith('.')]
+				hours.sort()
+				previous_hour_segments = None
+				all_hour_holes = {}
+				all_hour_partials = {}
+				for hour in hours:
+					if self.stopping.is_set():
+						break
+					self.logger.info('Checking {}/{}'.format(quality, hour))
+
+					# based on common.segments.best_segments_by_start
+					# but more complicated to capture more detailed metrics
+					hour_path = os.path.join(self.base_dir, self.channel, quality, hour)
+					segment_names = [name for name in os.listdir(hour_path) if not name.startswith('.')]
+					segment_names.sort()
+					parsed = []
+					bad_segment_count = 0
+					for name in segment_names:
+						try:
+							parsed.append(common.parse_segment_path(os.path.join(hour_path, name)))
+						except ValueError as e:
+							self.logger.warn(e)
+							bad_segment_count += 1
+
+					full_segment_count = 0
+					partial_segment_count = 0
+					full_segment_duration = datetime.timedelta()
+					partial_segment_duration = datetime.timedelta()
+					full_overlaps = 0
+					full_overlap_duration = datetime.timedelta()
+					partial_overlaps = 0
+					partial_overlap_duration = datetime.timedelta()
+					best_segments = []
+					holes = []
+					editable_holes = []
+					previous = None
+					previous_editable = None
+					coverage = datetime.timedelta()
+					editable_coverage = datetime.timedelta()
+					only_partials = []
+
+					# loop over all start times
+					# first select the best segment for a start time
+					# then update coverage
+					for start_time, segments in itertools.groupby(parsed, key=lambda segment: segment.start):
+						full_segments = []
+						partial_segments = []
+						for segment in segments:
+							if segment.type == 'full':
+								full_segments.append(segment)
+								full_segment_count += 1
+								full_segment_duration += segment.duration
+							elif segment.type == 'partial':
+								partial_segments.append(segment)
+								partial_segment_count += 1
+								partial_segment_duration += segment.duration
+						if full_segments:
+							if len(full_segments) == 1:
```

Comment:
you can remove this if, because the code under the `else` will still do the right thing when length == 1. ie. it'll sort, pick the last (only) one, and then the for loop will loop 0 times since there's no other things in the list.